### PR TITLE
Fixes #26191 - Find existing system by UUID during registration

### PR DESCRIPTION
**Pull request by jturel**
The original report observed a '409 Conflict' error when registering the same (previously registered) client repeatedly with a script like this:

```
#!/usr/bin/env bash
# Run this as "h=<server hostname> ./register-systems.sh"
set -x

u=${u:-admin}
p=${p:-changeme}
h=${h:-$(hostname)}
o=${o:-Default_Organization}
e=${e:-Library}

for i in $(seq 1 2000); do
        echo $i
        subscription-manager clean
        hostnamectl set-hostname katello-client-$i.example.com
        subscription-manager register --username=$u --password=$p --environment="$e" --serverurl=http://$h --auto-attach --org="$o"
        res=$?
        echo $res
done
```

The reason for the error is because even though Katello sees the registration as net-new, Candlepin is able to find the existing host profile by the dmi.system.uuid fact which doesn't change across registrations. This was changed in Candlepin 2.5

When taking that (already existing) UUID from Candlepin and creating the consumer in Pulp the '409 Conflict' is raised since the consumer already exists there.

This change teaches Katello to look for existing system profiles based on `dmi.system.uuid` in addition to current hostname. It will break the above script in the sense that it won't create 2000 profiles any more - it'll unregister and re-register the system (in backend terms, not Katello proper). In order to create multiple profiles the `dmi.system.uuid` fact will need to be overridden on each registration

open questions:

- [ ] should we use the 'raise an error when multiple profiles are found' approach I used here, or something else?
- [ ] if we keep raising an error, how good is the messaging?
- [ ] currently the system in Katello will not get the updated hostname per the above script when a profile is reused. should I change it to take the new hostname?

### CODE REVIEW

File: test/models/host/subscription_facet_test.rb [hunk 1]
```diff
@@ -220,6 +220,45 @@ def test_find_host
...
+      host2 = FactoryBot.create(:host, organization: org)
+
+      fact_name = FactName.create(name: 'dmi::system::uuid')
+      FactValue.create(value: "existing_system_uuid", host: host2, fact_name: fact_name)
+
+      facts = {'dmi.system.uuid' => 'existing_system_uuid', 'network.hostname' => host.name}
+
+      # if we get two matches, raise an error
+      error = assert_raises(RuntimeError) { Katello::Host::SubscriptionFacet.find_host(facts, org) }
+      assert_match(/since multiple were found/, error.message)
```
Comment (jturel) [inline test/models/host/subscription_facet_test.rb]: Do we want to raise an error here or do something else like choose the most recently created profile, see what candlepin knows about it, something else?

File: app/models/katello/host/subscription_facet.rb [hunk 2]
```diff
@@ -242,14 +242,21 @@ def remove_subscriptions(pools_with_quantities)
...
+        return nil if hosts.empty?
 
-        return nil if hosts.empty? #no host exists
         if hosts.where("organization_id = #{organization.id} OR organization_id is NULL").empty? #not in the correct org
           #TODO: http://projects.theforeman.org/issues/11532
           fail _("Host with name %{host_name} is currently registered to a different org, please migrate host to %{org_name}.") %
                    {:org_name => organization.name, :host_name => host_name }
         end
+
+        fail _("Could not associate this host to an existing profile since multiple were found. Please remove outdated content hosts.") if hosts.size > 1
```
Comment (jturel) [inline app/models/katello/host/subscription_facet.rb]: If we keep the failure mode here, would like some opinions on the messaging!

Comment (cfouant) [reply app/models/katello/host/subscription_facet.rb]: Personally, I do think the failure message is appropriate to keep. I think this one should be reworded for clarity? "Multiple profiles exist for this host. Please remove outdated content hosts." Or something along those lines? Just needs to be more succinct methinks.

Comment (jturel) [reply app/models/katello/host/subscription_facet.rb]: Yeah, mine was super wordy. I like yours!

**Comment by jturel** [comment]
@jlsherrill what are your thoughts on updating the hostname when cannibalizing an existing record in the db? asking since we've talked about this change previously.

File: app/models/katello/host/subscription_facet.rb (hunk 2 above)
Comment (jlsherrill) [reply app/models/katello/host/subscription_facet.rb]: could you include something like the host ids or hostnames, to make it easier on the user to identify?   
Hostnames might leak information, but a) the user has to be authed with at least an activation key, b) the only way they would conflict is if it already had the same hostname, or had the UUID of the existing system which seems strong enough to depend on.  So i'm leaning towards names

**Comment by jlsherrill** [comment]
@jturel per irc discussion, i think we should just leave it as is

File: app/models/katello/host/subscription_facet.rb (hunk 2 above)
Comment (jturel) [reply app/models/katello/host/subscription_facet.rb]: @cfouant @jlsherrill 

Updated the message to include matching hostnames - looks something like this:

```
Multiple profiles found. Consider removing host12, host13 which match this host.
```

**Comment by jturel** [comment]
[test katello]

**Comment by jturel** [comment]
[test katello]

**Comment by jturel** [comment]
[test katello]

**Comment by jturel** [comment]
[test katello]

**Comment by jturel** [comment]
@cfouant @jlsherrill any more input on this one?

**Pull request merged**